### PR TITLE
local-gadget: fix tracing new crio containers

### LINKED ISSUE #1018: local-gadget: Trace events not shown for new cri-o containers started after trace command
Currently, if you start a trace on a node e.g `local-gadget trace bind`, all the cri-o containers started after the command aren't traced and no trace events are printed. Terminating and starting the `local-gadget` again fixes the issue. This happens only on `CRI-O` based environment.

## Impact

We will miss events from the containers started after the trace. It only impacts CRI-O based environments and local-gadget.

## Environment and steps to reproduce

1. Setup a minikube with cri-o using: `make -C integration/local-gadget setup-cri-o`
2. Ensure you don't hit [CAP apply](https://github.com/kubernetes/minikube/issues/12928#issuecomment-1245458044) issue. 
3. Copy the local-gadget to the container using: `docker cp local-gadget minikube-cri-o:/bin/local-gadget`
4. Start the the tracing using: `docker exec -it minikube-cri-o local-gadget trace bind`
5. In other terminal start the test container and try some socket binding
```
kubectl run test-pod --image busybox:latest sleep inf
kubectl exec -ti test-pod -- nc -l -p 4242 -w 1
```
6.  No events are shown by trace started by 4. 

## Expected behavior

Trace events should be printed.

## Additional information

Initial finding suggest adding runc path for cri-o to runcPaths might work but it needs to be investigated.

### PR DESCRIPTION
**Pull request by mqasimsarfraz**
`runcfanotify` doesn't work for `cri-o` containers. The idea here is to adopt it to support tracing `cri-o` containers. 

Fixes #1018
Related #1028

### Idea
During `runc` notification container creation is blocked on `pidfile` so enriching the container using the container runtime won't work because we hit `Runtime enricher (cri-o): failed to get container: container "2899fb3ce898a2ed1d4426b0cb6df81ee72cd630ce0fb9e10eb53f04d5eb4e9b" not found`. In order to avoid this we are using annotations fetched via `OCIConfig` to enrich containers. 

### Usage
```
$ docker exec -it minikube-cri-o bash
$ local-gadget list-containers --watch 
# Create a container in another terminal e.g kubectl run test-pod --image busybox:latest sleep inf
...
cri-o      a0ddd82147f5536d38c449f39edff4151ae0aa39b9653045449302915c85… test-pod
...
```
JSON mode should also work here. Before this PR this will only works for `docker/containerd` but now it should work for `cri-o` as well.

### Future work
Consider using annotations for other runtimes or other information like `cgroups` etc. Having said that, `containerd` has some annotations but `docker` has none.

### CODE REVIEW

File: pkg/container-collection/options.go [hunk 1]
```diff
@@ -585,6 +581,30 @@ func WithLinuxNamespaceEnrichment() ContainerCollectionOption {
 	}
 }
 
+// WithOCIConfigEnrichment enriches container using provided OCI config
+func WithOCIConfigEnrichment() ContainerCollectionOption {
+	return func(cc *ContainerCollection) error {
+		cc.containerEnrichers = append(cc.containerEnrichers, func(container *Container) bool {
+			if container.OciConfig != nil {
```
Comment (blanquicet) [inline pkg/container-collection/options.go:588]: What about using early return?

```golang
	if container.OciConfig == nil {
		return true
	}
```

File: pkg/container-collection/options.go [hunk 2]
```diff
@@ -585,6 +581,30 @@ func WithLinuxNamespaceEnrichment() ContainerCollectionOption {
...
+				if cn, ok := container.OciConfig.Annotations[crioContainerNameAnnotation]; ok {
+					container.Name = cn
+				}
+				if pn, ok := container.OciConfig.Annotations[crioPodNameAnnotation]; ok {
+					container.Podname = pn
+				}
+				if ns, ok := container.OciConfig.Annotations[crioPodNamespaceAnnotation]; ok {
+					container.Namespace = ns
+				}
+			}
```
Comment (blanquicet) [inline pkg/container-collection/options.go:601]: It should also enrich with the Pod UID.

File: pkg/container-collection/options.go (hunk 1 above)
Comment (blanquicet) [inline pkg/container-collection/options.go:588]: Shouldn't we verify that container is not already enriched as we are doing [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/f6b8c02bd5d1515e022fbf8cec220e7fad09d57f/pkg/container-collection/options.go#L67-L71)? BTW, we should also verify if the Pod UID and the Runtime.

File: cmd/local-gadget/containers/containers.go [hunk 3]
```diff
@@ -44,33 +48,69 @@ func NewListContainersCmd() *cobra.Command {
...
-
 			return nil
 		},
 	}
 
+	cmd.Flags().BoolVarP(
+		&optionWatch,
+		"watch", "w",
+		false,
+		"After listing the containers, watch for new containers")
```
Comment (blanquicet) [inline cmd/local-gadget/containers/containers.go:86]: Please add a new test in `integration/local-gadget/list_containers_test.go` to verify this feature.

**Review by blanquicet — COMMENTED** [review]
Thanks for working on this! These are just some comments, but it indeed fixed the issue.

File: pkg/container-collection/annotations.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// limitations under the License.
+
+package containercollection
+
+const (
+	// cri-o container annotations to get container information
+	crioContainerManagerAnnotation = "io.container.manager"
+	crioPodNameAnnotation          = "io.kubernetes.pod.name"
+	crioPodNamespaceAnnotation     = "io.kubernetes.pod.namespace"
+	crioContainerNameAnnotation    = "io.kubernetes.container.name"
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply pkg/container-collection/annotations.go:22]: I checked what containerd was doing and found out they have annotations named "io.kubernetes.cri.container-name", "io.kubernetes.cri.sandbox-namespace" so decided to put the crio prefix for our clarity.

File: pkg/container-collection/options.go [hunk 5]
```diff
@@ -585,6 +581,34 @@ func WithLinuxNamespaceEnrichment() ContainerCollectionOption {
...
 
+// WithOCIConfigEnrichment enriches container using provided OCI config
+func WithOCIConfigEnrichment() ContainerCollectionOption {
+	return func(cc *ContainerCollection) error {
+		cc.containerEnrichers = append(cc.containerEnrichers, func(container *Container) bool {
+			if container.OciConfig != nil {
+				// skip for pause container
+				if ct, ok := container.OciConfig.Annotations[crioContainerType]; ok && ct == "sandbox" {
+					return false
+				}
```
Comment (blanquicet) [inline pkg/container-collection/options.go:592]: I don't think we should skip pause containers, as local-gadget users may also want to debug those. In fact, the ContainerRuntime enricher doesn't drop them.

File: pkg/container-collection/annotations.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/container-collection/annotations.go:22]: Oh, then the name is ok. Thanks for the clarification.

File: pkg/container-collection/options.go (hunk 5 above)
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go:592]: Got it. I realized that pause containers were being printed with `list-containers --watch` I feel like since we aren't listing them with non-watched listing we should skip them during watch as well. What do you think?

Comment (blanquicet) [reply pkg/container-collection/options.go:592]: Mmmm I remember to have seen the pause containers only for Docker on both cases, watched and non-watched. What runtime were you using when you saw this behaviour?

**Comment by mauriciovasquezbernal** [comment]
> During runc notification container creation is blocked on pidfile so enriching the container using the container runtime won't work because we hit Runtime enricher (cri-o): failed to get container: container "2899fb3ce898a2ed1d4426b0cb6df81ee72cd630ce0fb9e10eb53f04d5eb4e9b" not found. In order to avoid this we are using annotations fetched via OCIConfig to enrich containers.

I don't understand what is the problem here. Why does it work with other runtimes? 

Btw, this commit also fixes the issue with inspektor gadget but only when `--hook-mode=fanotify` is used. I'll investigate why CRI-O hooks aren't working fine.

**Comment by blanquicet** [comment]
> > During runc notification container creation is blocked on pidfile so enriching the container using the container runtime won't work because we hit Runtime enricher (cri-o): failed to get container: container "2899fb3ce898a2ed1d4426b0cb6df81ee72cd630ce0fb9e10eb53f04d5eb4e9b" not found. In order to avoid this we are using annotations fetched via OCIConfig to enrich containers.
> 
> I don't understand what is the problem here. Why does it work with other runtimes?
> 
From what I understood, unlike docker and containerd, when runc-fanotify notifies that a container is being created, CRI-O has not exposed the container through the CRI yet. Therefore, the ContainerRuntime enricher fails to gather the information needed to enrich the internal container structure.

**Comment by blanquicet** [comment]
> > > During runc notification container creation is blocked on pidfile so enriching the container using the container runtime won't work because we hit Runtime enricher (cri-o): failed to get container: container "2899fb3ce898a2ed1d4426b0cb6df81ee72cd630ce0fb9e10eb53f04d5eb4e9b" not found. In order to avoid this we are using annotations fetched via OCIConfig to enrich containers.
> > 
> > 
> > I don't understand what is the problem here. Why does it work with other runtimes?
> 
> From what I understood, unlike docker and containerd, when runc-fanotify notifies that a container is being created, CRI-O has not exposed the container through the CRI yet. Therefore, the ContainerRuntime enricher fails to gather the information needed to enrich the internal container structure.
> 
> Before introducing this patch, I think we should try to understand why it behaves differently from other runtimes.

BTW, I didn't see the notification from runc-fanotify about the pause container when using CRI-O. Is there something wrong with my setup?

**Comment by mqasimsarfraz** [comment]
> BTW, I didn't see the notification from runc-fanotify about the pause container when using CRI-O. Is there something wrong with my setup?

I only see them for `kvm` driver so need to investigate that is the difference between the 2 setups. I am setting it up using: `minikube start -p minikube-kvm --driver=kvm --iso-url=https://storage.googleapis.com/minikube-builds/iso/15232/minikube-v1.27.0-1666976405-15232-amd64.iso --container-runtime=cri-o`.

File: pkg/container-collection/options.go (hunk 5 above)
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go:592]: CRI-O on kvm driver. I am bit confused regrading pause container and pod sandbox because the notifications I was seeing are related to pod sandbox. We don't need to show them to users. Right?

**Comment by mqasimsarfraz** [comment]
> > I don't understand what is the problem here. Why does it work with other runtimes?
> 
> From what I understood, unlike docker and containerd, when runc-fanotify notifies that a container is being created, CRI-O has not exposed the container through the CRI yet. Therefore, the ContainerRuntime enricher fails to gather the information needed to enrich the internal container structure.
> 
+1 and just adding to that you can verify this behavior on your setup by starting a verbose trace from `main` e.g `local-gadget trace bind -v -r cri-o`
 and try starting a pod in another terminal. You would see the "not found" error thrown by CRI client (even if you trim the `userdata` in path). Same workflow works for `containerd`. 
> Before introducing this patch, I think we should try to understand why it behaves differently from other runtimes.

From what I understood, since we are blocking the `pidfile` to capture the first event. `conmon` gets blocked till we have enriched and then only the container is available via CRI. But definitely makes to confirm this.

File: pkg/container-collection/options.go (hunk 5 above)
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go:592]: I think I have an idea what confused me. We aren't listing the pod sandboxes in the [CRI client](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/container-utils/cri/cri.go#L80) so they are dropped during enrichment even for `contrainerd`. If we decided to handle them here we should  adopt the cri client to handle them as well.

**Comment by blanquicet** [comment]
> > Before introducing this patch, I think we should try to understand why it behaves differently from other runtimes.
> 
> From what I understood, since we are blocking the `pidfile` to capture the first event. `conmon` gets blocked till we have enriched and then only the container is available via CRI. But definitely makes to confirm this.
>
Yep, but why it doesn't happen with other runtimes?

File: pkg/container-collection/options.go [hunk 6]
```diff
@@ -585,6 +581,34 @@ func WithLinuxNamespaceEnrichment() ContainerCollectionOption {
 	}
 }
 
+// WithOCIConfigEnrichment enriches container using provided OCI config
+func WithOCIConfigEnrichment() ContainerCollectionOption {
```
Comment (alban) [inline pkg/container-collection/options.go:585]: Since the name of this function is not specific to cri-o, should you also enrich from containerd annotations?

File: pkg/container-collection/options.go (hunk 5 above)
Comment (blanquicet) [reply pkg/container-collection/options.go:592]: Ahh, you are right. We should use [`ListPodSandbox`](https://pkg.go.dev/k8s.io/cri-api@v0.20.6/pkg/apis/runtime/v1alpha2#RuntimeServiceClient.ListPodSandbox) to get the pause containers. While using the Docker API Engine, we get all of them with [the same API](https://pkg.go.dev/github.com/docker/docker@v20.10.8+incompatible/client#Client.ContainerList).

I propose to continue ignoring the pause containers for CRI-O and containerd in this PR, and manage that [issue](https://github.com/inspektor-gadget/inspektor-gadget/issues/1095) in another one.

File: pkg/container-collection/options.go (hunk 2 above)
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go:605]: Yes it is available but just for crio via annotation: `io.kubernetes.pod.uid`. I will go ahead and add it.

**Comment by mqasimsarfraz** [comment]
> Yep but, why doesn't happen with other runtimes?

I added some delay to `GetContainer` to compare the state between `containerd` and `cri-o` while blocked and found not that in case of `containerd` container is available via CRI right after `runc create` but in case of `cri-o` it is only available after the `conmon` communicates the container_pid to the `cri-o` server. This seem to be the limitation of the design since `containerd` takes care of creating container by itself, it has more awareness of container states while `cri-o` only knows about container state once it [hears back](https://github.com/containers/conmon/blob/main/src/conmon.c#L396) from `conmon`, side effect of delegating the creation process.

```
	container_pid = atoi(contents);
	ndebugf("container PID: %d", container_pid);

	g_hash_table_insert(pid_to_handler, (pid_t *)&container_pid, container_exit_cb);

	/* Send the container pid back to parent
	 * Only send this pid back if we are using the current exec API. Old consumers expect
	 * conmon to only send one value down this pipe, which will later be the exit code
	 * Thus, if we are legacy and we are exec, skip this write.
	 */
	if ((opt_api_version >= 1 || !opt_exec) && sync_pipe_fd >= 0)
		write_sync_fd(sync_pipe_fd, container_pid, NULL);
```
or with the logs:

```
Nov 07 17:04:01 minikube-cri-o crio[423]: time="2022-11-07 17:04:01.980982415Z" level=info msg="Creating container: default/test-pod-25833/test-pod-25833" id=b90317f1-aeb6-46c0-84a8-3bec0a399770 name=/runtime.v1.RuntimeService/CreateContainer
Nov 07 17:04:01 minikube-cri-o crio[423]: time="2022-11-07 17:04:01.981058169Z" level=warning msg="Allowed annotations are specified for workload []"
Nov 07 17:04:02 minikube-cri-o crio[423]: time="2022-11-07 17:04:02.124023823Z" level=warning msg="Unable to find container ID ef99145814d2ae9e3e0ede9fd58cf9eb7f06985756e0322070c81da24f2c611e" id=a163e4e8-66a8-4b30-8581-b45dbd57f1c6 name=/runtime.v1alpha2.RuntimeService/ListContainers
Nov 07 17:04:02 minikube-cri-o crio[423]: time="2022-11-07 17:04:02.161124862Z" level=info msg="Created container ef99145814d2ae9e3e0ede9fd58cf9eb7f06985756e0322070c81da24f2c611e: default/test-pod-25833/test-pod-25833" id=b90317f1-aeb6-46c0-84a8-3bec0a399770 name=/runtime.v1.RuntimeService/CreateContainer
Nov 07 17:04:02 minikube-cri-o crio[423]: time="2022-11-07 17:04:02.161552331Z" level=info msg="Starting container: ef99145814d2ae9e3e0ede9fd58cf9eb7f06985756e0322070c81da24f2c611e" id=333c6b4f-ae07-4ff4-b5a1-188a5ac85ed7 name=/runtime.v1.RuntimeService/StartContainer
Nov 07 17:04:02 minikube-cri-o crio[423]: time="2022-11-07 17:04:02.197041419Z" level=info msg="Started container" PID=2765 containerID=ef99145814d2ae9e3e0ede9fd58cf9eb7f06985756e0322070c81da24f2c611e description=default/test-pod-25833/test-pod-25833 id=333c6b4f-ae07-4ff4-b5a1-188a5ac85ed7 name=/runtime.v1.RuntimeService/StartContainer sandboxID=d86c2b34c8ca9c838b9505b41c69c18dac51ffa12d4924aa109e2f3643cc6265
```
[Creation](https://github.com/cri-o/cri-o/blob/main/server/container_create.go#L443) only resumes after `enrichers` are done and pidfile is free. With this we know why we need this change for `cri-o` and not for `containerd`.

**Comment by mqasimsarfraz** [comment]
I have updated the PR to handle the comments. Would be great to get another review!

File: integration/local-gadget/list_containers_test.go [hunk 7]
```diff
@@ -89,41 +75,22 @@ func TestListContainers(t *testing.T) {
 
 func TestListSingleContainer(t *testing.T) {
 	t.Parallel()
-	prefix := "test-list-single"
-	po := fmt.Sprintf("%s-pod", prefix)
-	cn := fmt.Sprintf("%s-container", prefix)
-	ns := GenerateTestNamespaceName(fmt.Sprintf("%s-namespace", prefix))
+	cn := "test-list-single"
```
Comment (blanquicet) [inline integration/local-gadget/list_containers_test.go:78]: nit: I would use a diff name for the pod.

File: integration/local-gadget/list_containers_test.go [hunk 8]
```diff
@@ -143,39 +110,68 @@ spec:
...
-			ExpectedRegexp: fmt.Sprintf("pod/%s created", po),
-		},
-		{
-			Name:           "WaitForTestListSinglePod",
-			Cmd:            fmt.Sprintf("kubectl wait pod --for condition=ready -n %s %s", ns, po),
-			ExpectedString: fmt.Sprintf("pod/%s condition met\n", po),
-		},
-		listContainersCmd,
+		watchContainersCmd,
+		PodCommand(cn, "busybox", ns, `["sleep", "inf"]`, ""),
```
Comment (blanquicet) [inline integration/local-gadget/list_containers_test.go:173]: What can we do to ensure `local-gadget` is already running when we start the pod? Maybe a simple `sleep` 😅? I mean, from the output, it is difficult to understand whether `local-gadget` captured it from [the already running containers](https://github.com/inspektor-gadget/inspektor-gadget/blob/5a9c080fc0ac14c63694d91e825b5d6566661a3d/cmd/local-gadget/containers/containers.go#L77-L80), or [the `EventTypeAddContainer` event](https://github.com/inspektor-gadget/inspektor-gadget/blob/5a9c080fc0ac14c63694d91e825b5d6566661a3d/cmd/local-gadget/containers/containers.go#L65-L68). So, we need another way to ensure that, don't you think?

File: pkg/container-collection/options.go [hunk 9]
```diff
@@ -585,6 +582,65 @@ func WithLinuxNamespaceEnrichment() ContainerCollectionOption {
...
+					container.Podname = pn
+				}
+				if ns, ok := container.OciConfig.Annotations[containerdPodNamespaceAnnotation]; ok {
+					container.Namespace = ns
+				}
+				container.Runtime = runtimeclient.ContainerdName
+				return true
+			}
+
+			return true
```
Comment (blanquicet) [inline pkg/container-collection/options.go:638]: I thought something like this could look better, but I am not that sure anymore:

```suggestion
			var containerTypeAnnotation, containerNameAnnotation, podNameAnnotation, podNamespaceAnnotation string

			if cm := container.OciConfig.Annotations[crioContainerManagerAnnotation]; cm == runtimeclient.CrioName {
				container.Runtime = runtimeclient.CrioName

				containerTypeAnnotation = crioContainerTypeAnnotation
				containerNameAnnotation = crioContainerNameAnnotation
				podNameAnnotation = crioPodNameAnnotation
				podNamespaceAnnotation = crioPodNamespaceAnnotation

				if pUID, ok := container.OciConfig.Annotations[crioPodUIDAnnotation]; ok {
					container.PodUID = pUID
				}
			} else if _, isContainerd := container.OciConfig.Annotations[containerdContainerTypeAnnotation]; isContainerd {
				container.Runtime = runtimeclient.ContainerdName

				containerTypeAnnotation = containerdContainerTypeAnnotation
				containerNameAnnotation = containerdContainerNameAnnotation
				podNameAnnotation = containerdPodNameAnnotation
				podNamespaceAnnotation = containerdPodNamespaceAnnotation
			} else {
				return true
			}

			// TODO: handle this once we support pod sandboxes via WithContainerRuntimeEnrichment
			// Issue: https://github.com/inspektor-gadget/inspektor-gadget/issues/1095
			if ct := container.OciConfig.Annotations[containerTypeAnnotation]; ct == containerTypeSandbox {
				return false
			}

			if cn, ok := container.OciConfig.Annotations[containerNameAnnotation]; ok {
				container.Name = cn
			}
			if pn, ok := container.OciConfig.Annotations[podNameAnnotation]; ok {
				container.Podname = pn
			}
			if ns, ok := container.OciConfig.Annotations[podNamespaceAnnotation]; ok {
				container.Namespace = ns
			}

			return true
```

**Review by blanquicet — COMMENTED** [review]
I will test it tomorrow, but I think it is ready to go. Thanks for managing all the comments and investigating the root cause. Awesome work!

File: integration/local-gadget/list_containers_test.go (hunk 8 above)
Comment (mqasimsarfraz) [reply integration/local-gadget/list_containers_test.go:173]: We definitely need to ensure this. I hit it couple of times while running tests. For now we do a good old `sleep` and look into other ways latter (if there are any) ?

File: cmd/local-gadget/containers/containers.go [hunk 10]
```diff
@@ -44,33 +50,73 @@ func NewListContainersCmd() *cobra.Command {
 				return commonutils.WrapInErrParserCreate(err)
 			}
 
-			containers := localGadgetManager.GetContainersBySelector(&containercollection.ContainerSelector{
-				Name: commonFlags.Containername,
-			})
+			w := bufio.NewWriter(os.Stdout)
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply cmd/local-gadget/containers/containers.go:55]: OK, but please update the [documentation](https://github.com/inspektor-gadget/inspektor-gadget/blob/a6cd04153237b493e7b2c79627eef33d38d66420/docs/local-gadget.md?plain=1#L119-L137).

**Review by blanquicet — APPROVED** [review]
LGTM after updating the documentation. Thanks!

**Comment by mqasimsarfraz** [comment]
> LGTM after updating the documentation. Thanks!

Thanks a lot with all the reviews. Super helpful. Once `main` is fixed, let's us get a green run and merge this?

**Comment by blanquicet** [comment]
> > LGTM after updating the documentation. Thanks!
> 
> Thanks a lot with all the reviews. Super helpful. Once `main` is fixed, let's us get a green run and merge this?

Yep, let's fix first the CI, so that we can verify this PR is green before merging.

File: cmd/local-gadget/containers/containers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply cmd/local-gadget/containers/containers.go:55]: > We changed from writing a list of containers to printing one container at a time

Why to change that? As a user I'd expect it to be a list (when not using the -w option). 

Please notice that `kubectl get pods -o json -w` prints a json object that has a list inside, so it's not exactly the same case.

File: pkg/local-gadget-manager/local-gadget-manager.go [hunk 11]
```diff
@@ -348,6 +348,7 @@ func NewManager(runtimes []*containerutils.RuntimeConfig) (*LocalGadgetManager,
 
 	err = l.ContainerCollection.Initialize(
 		containercollection.WithPubSub(containerEventFuncs...),
+		containercollection.WithOCIConfigEnrichment(),
```
Comment (mauriciovasquezbernal) [inline pkg/local-gadget-manager/local-gadget-manager.go:351]: I think it should also be added to `pkg/gadgettracermanager/gadgettracermanager.go`, otherwise this fix won't be available to Inspektor Gadget in Kubernetes.

File: integration/command.go [hunk 12]
```diff
@@ -598,3 +598,11 @@ func WaitUntilPodReadyCommand(namespace string, podname string) *Command {
 func WaitUntilTestPodReadyCommand(namespace string) *Command {
 	return WaitUntilPodReadyCommand(namespace, "test-pod")
 }
+
+// WaitForSecondsCommand returns a Command which waits for given seconds
+func WaitForSecondsCommand(seconds int) *Command {
```
Comment (mauriciovasquezbernal) [inline integration/command.go:603]: Use `SleepForSecondsCommand` to be more explicit?

**Review by mauriciovasquezbernal — COMMENTED** [review]
Some comments about the json output format.

File: cmd/local-gadget/containers/containers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (blanquicet) [reply cmd/local-gadget/containers/containers.go:55]: > As a user I'd expect it to be a list (when not using the -w option).
>
I'm not sure. Considering how other gadgets works, I wouldn't. In addition, I will definitely not expect the output format changes when using this kind of option (`--watch`). IMO, it's annoying for people to parse the output, don't you think?

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply cmd/local-gadget/containers/containers.go:55]: > it's annoying for people parsing the output to have to manage different formats of the same command, don't you think?

I agree, but I also think it's annoying that the output of a `list-containers` command is not something I can parse as a json object (with this PR):

```bash
$ sudo ./local-gadget list-containers -o json | jsonlint-php 
WARN[0000] Runtime enricher (cri-o): couldn't get current containers 
Parse error on line 14:
... "local-registry"}{  "runtime": "dock
--------------------^
Expected one of: 'EOF', '}', ',', ']' (stdin)
```

For gadgets that print multiple times to the terminal (tracers, top, audit, ...) what we do is that **each** event is printed as a json object to the terminal in a new line, then the client can consume each line as a valid json. 

To summarize:
1. Gadgets that print multiple objects a once and then return (snapshot) print a valid json array.
2. Gadgets that print multiple events print an event per line using json. 

The point is that `list-containers` falls on the first category but `list-containers -w` falls on the second one. 

Are my concerns clear?

[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply cmd/local-gadget/containers/containers.go:55]: Okay so I see the point here. Just to summarize my understanding: for most of the gadgets we are **waiting** for the json events so it is okay to stream a single json object at a time but for "print and exit" commands we should output a valid json?  Having said that, it then makes sense to revert to printing json list for `list-containers` and printing stream of json objects for `list-containers -w` ?

[the remark this answers is not in the document]
Comment (blanquicet) [reply cmd/local-gadget/containers/containers.go:55]: > Are my concerns clear?
>
Yep, they are.

> The point is that `list-containers` falls on the first category but `list-containers -w` falls on the second one. 
>
However, I still find this pretty strange. I can't recall any other CLI doing something similar. 

Anyway, let's move this PR on. I am OK printing a JSON array for `list-containers` and a stream of JSON objects for `list-containers -w`.

[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply cmd/local-gadget/containers/containers.go:55]: > > The point is that `list-containers` falls on the first category but `list-containers -w` falls on the second one.
> 
> However, I still find this pretty strange. I can't recall any other CLI doing something similar.

I also thought the same but it seem `kubectl` falls in this category with different outputs.
```
→ kubectl get pods -o json 
{
    "apiVersion": "v1",
    "items": [
       ...
    ],
    "kind": "List",
    "metadata": {
        "resourceVersion": ""
    }
}

```
and 

```
→ kubectl get pods -o json -w
{
    "apiVersion": "v1",
    "kind": "Pod",
    ...
}
```
Anyhow, let me update the PR.

File: pkg/local-gadget-manager/local-gadget-manager.go (hunk 11 above)
Comment (mqasimsarfraz) [reply pkg/local-gadget-manager/local-gadget-manager.go:351]: True. But I prefer not to hanldle it in the same PR. I wanted to have look into `crio` hook mode and do some testing for IG with this.  I opened the [issue](https://github.com/inspektor-gadget/inspektor-gadget/issues/1142) to track this. What do you think?

File: pkg/container-collection/annotations.go [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+// limitations under the License.
+
+package containercollection
+
+const (
+	// container types used by the runtimes
+	containerTypeSandbox   = "sandbox"
+	containerTypeContainer = "container"
+
+	// cri-o container annotations to get container information
```
Comment (alban) [inline pkg/container-collection/annotations.go:22]: Could you add a comment saying where it comes from? I can see some of them at:

https://github.com/containers/podman/blob/main/pkg/annotations/annotations.go
https://github.com/containerd/containerd/blob/master/pkg/cri/annotations/annotations.go

But not all of them.

File: pkg/container-collection/options.go [hunk 14]
```diff
@@ -585,6 +582,54 @@ func WithLinuxNamespaceEnrichment() ContainerCollectionOption {
...
+// WithOCIConfigEnrichment enriches container using provided OCI config
+func WithOCIConfigEnrichment() ContainerCollectionOption {
+	return func(cc *ContainerCollection) error {
+		cc.containerEnrichers = append(cc.containerEnrichers, func(container *Container) bool {
+			if container.OciConfig == nil || container.IsEnriched() {
+				return true
+			}
+
+			var containerTypeAnnotation, containerNameAnnotation, podNameAnnotation, podNamespaceAnnotation string
+			if cm := container.OciConfig.Annotations[crioContainerManagerAnnotation]; cm == runtimeclient.CrioName {
```
[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go:594]: > I wonder if it could even be a separate package, so it could be reused by [seccompagent's kuberesolver.go#L30-L41](https://github.com/kinvolk/seccompagent/blob/main/pkg/kuberesolver/kuberesolver.go#L30-L41) without importing all container-collection but that would complicate `enrichFromAnnotations()`.

True, that would be nice. How about we return a custom `struct` with all the currently supported values in a standalone package? Later on it can be mapped to `containercollection.Container` in `enrichFromAnnotations` ?

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM.

File: pkg/container-collection/options.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (alban) [reply pkg/container-collection/options.go:594]: That sounds good. Could this be done in a different PR? So that discussions on the details don't postpone this improvement on crio support.

[the remark this answers is not in the document]
Comment (mqasimsarfraz) [reply pkg/container-collection/options.go]: I opened the [PR](https://github.com/inspektor-gadget/inspektor-gadget/pull/1147) with the changes. Should we go ahead and merge this one?

**Comment by mqasimsarfraz** [comment]
Thanks for all the reviews! I will go ahead and merge this!
